Use XDG config directory

klmr opened the pull request:
It would be great if the default lintr config would be preferentially looked up in the appropriate [XDG base configuration directory](https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html) instead of the user home directory. That is, if lintr defaulted to loading `${XDG_CONFIG_HOME-$HOME/.config}/lintr` (note lack of leading dot in the filename!) and only if that doesn’t exist, checked for `$HOME/.lintr`. This avoids cluttering the home directory with config files.

Of course this can already be achieved by appropriately setting the `lintr.linter_file` option. But defaults matter.

This PR implements the necessary change, and stays 100% backwards compatible. Note that this causes the cyclomatic complexity of the `find_config` function to exceed the linting threshold. I’m open for suggestions of how to change this — but I’m tempted to call it a false positive (and to add an appropriate `nolint` comment).

gaborcsardi commented:
How about the rappdirs package?

klmr commented:
Good idea! However, if I understand correctly rappdirs is particularly geared towards GUI applications, rather than cross-platform tools and/or code library packages. In particular, on macOS, `rappdirs::user_config_dir('lintr')` returns `/Users/‹user›/Library/Application Support/lintr`, which is generally *not* where a command line tool/library configuration would be expected to reside. Apple would probably disagree, but the overwhelming convention for such tools on macOS is to either follow XDG or to dump everything in the home dir.

jimhester commented:
I think using rappdirs is the best solution if we want to go down this route.

AshesITR commented:
Meanwhile, base R provides `tools::R_user_dir("lintr", "config")`, with `backports` providing this function to older R versions.

AshesITR commented:
@krlmlr Are you still interested in this? #1757 is about to change the config search behavior so now would be a good time to also include XDG_CONFIG_HOME if it still seems useful.

klmr commented:
@AshesITR (Wrong user tagged) Absolutely, and I had thought I had already submitted an updated PR. My bad! I’ll check the other PR and submit my changes ASAP!

AshesITR commented:
Thanks a ton and sorry for the wrong tag.
Maybe best to base against the new PR because it simplifies the code complexity by creating a candidate list.
This PR would then be as small as +1 line of code and some documentation around it.

klmr commented:
OK, I’ve added the necessary changes. This includes adding an environment variable to override the lintr config file location; I added this change because I think it belongs here: it is a widespread convention that the location of a configuration file can be overridden by an environment variable.

In the case of ‘lintr’ it made sense (I think) to make this environment variable act as the default value for `getOption('lintr.linter_file')`, rather than being yet another way of specifying the file location.

@AshesITR At your suggestion I rebased this PR onto #1757; however, this means that the outstanding changes to that PR will probably block this one.

AshesITR commented on R/settings_utils.R:
```diff
@@ -68,24 +68,26 @@ find_config <- function(filename) {
+    # cf: rstudio@bc9b6a5 SessionRSConnect.R#L32
+    file.path(Sys.getenv("HOME", unset = "~"), linter_file),
+    # Next check for a global config file
+    file.path(R_user_dir("lintr", which = "config"), "config")
```
Maybe use `lintr_file` as the filename here as well?

Also, from the earlier discussion I though that XDG_CONFIG_HOME might also be a useful search location, right?

klmr replied:
> Maybe use lintr_file as the filename here as well?

To clarify: you think the file path should be e.g. `~/.config/R/lintr/lintr_file` instead of `~/.config/R/lintr/config`? I’m happy with either but the latter is more common (e.g. Git has `~/.config/git/config`, etc).

> Also, from the earlier discussion I though that XDG_CONFIG_HOME might also be a useful search location, right?

`tool::R_user_dir` uses `XDG_CONFIG_HOME`, if set.

AshesITR replied:
I was thinking it should be `getOption("lintr.linter_file")` like with all other candidate paths.

klmr replied:
Hm, I disagree with this. Firstly, I don’t think there’s a benefit to making this configurable (the path *is* already configurable on multiple levels, including via `XDG_CONFIG_HOME` and `R_USER_CONFIG_DIR`).

And secondly, the filename given by `getOption("lintr.linter_file")` makes the file invisible on most platforms (leading dot by default). This is appropriate in the home directory or a project directory (…ish, to be honest). It is *not* appropriate or necessary in a dedicated config directory. Users (me included) might want to use an invisible filename for their projects (e.g. `.lintr`) but a regular file for the global configuration.

What do you think?

AshesITR replied:
Okay, works for me.

AshesITR commented:
Could you write some tests to see if the XDG config is pulled as expected?

klmr commented:
@AshesITR Apologies, I had completely missed that. Test added.

AshesITR commented:
LGTM, thanks! Waiting for #1757 to hit the finish line and then rebasing #460 again.

klmr commented:
@AshesITR I’ve rebased the PR against the merged `main` branch now that #1757 has been merged. Please have a look.

AshesITR commented:
Sorry for taking so long to review, I'll try to get to this soon.
